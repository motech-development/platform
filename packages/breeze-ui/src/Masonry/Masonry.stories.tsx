import { withA11y } from '@storybook/addon-a11y';
import { number, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import BaseStyles from '../BaseStyles/BaseStyles';
import Card from '../Card/Card';
import Typography from '../Typography/Typography';
import Masonry from './Masonry';

const stories = storiesOf('Masonry', module);

stories.addDecorator(withA11y);
stories.addDecorator(withKnobs);

stories.add('Basic masonry', () => (
  <>
    <BaseStyles />

    <Masonry
      xs={number('Columns in extra small viewport', 1)}
      sm={number('Columns in small viewport', 2)}
      md={number('Columns in medium viewport', 3)}
      lg={number('Columns in large viewport', 4)}
    >
      <Card>
        <div style={{ height: '200px' }}>
          <Typography rule component="h1" variant="h2">
            Outside of loop
          </Typography>
        </div>
      </Card>

      {[...Array(number('Items', 5))].map((_, i) => {
        const height = `${200 + Math.ceil(Math.random() * 300)}px`;

        return (
          // eslint-disable-next-line react/no-array-index-key
          <Card key={i}>
            <div style={{ height }}>
              <Typography rule component="h1" variant="h2">{`Item ${
                i + 1
              }`}</Typography>
            </div>
          </Card>
        );
      })}
    </Masonry>
  </>
));
