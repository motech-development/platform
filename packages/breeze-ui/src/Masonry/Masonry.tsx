import React, { FC, memo, ReactNode, useEffect, useState } from 'react';
import styled from 'styled-components';
import useWindowSize from '../hooks/useWindowSize';
import Col from '../Col/Col';
import Row from '../Row/Row';

interface IMasonryItemContainerProps {
  gutter: string;
}

const MasonryItem = styled.div<IMasonryItemContainerProps>`
  ${({ gutter }) => `
    margin-bottom: ${gutter};
  `}
`;

const getBreakpoint = () => {
  const lg = window.matchMedia('(min-width: 992px)');
  const md = window.matchMedia('(min-width: 768px)');
  const sm = window.matchMedia('(min-width: 576px)');

  if (lg.matches) {
    return 'lg';
  }

  if (md.matches) {
    return 'md';
  }

  if (sm.matches) {
    return 'sm';
  }

  return 'xs';
};

const calculateCols = (cols: number) => 12 / cols;

export interface IMasonryProps {
  children: ReactNode | ReactNode[];
  gutter?: string;
  lg: number;
  md: number;
  sm: number;
  xs: number;
}

const Masonry: FC<IMasonryProps> = ({
  children,
  gutter = '1rem',
  lg,
  md,
  sm,
  xs,
}) => {
  const breakpoints = {
    lg,
    md,
    sm,
    xs,
  };
  const [breakpoint, setBreakpoint] = useState(getBreakpoint);
  const columns = breakpoints[breakpoint];
  const { width } = useWindowSize();
  const cols: ReactNode[][] = Array.from(new Array(columns), () => []);
  const flat = Array.isArray(children) ? children.flat() : [children];

  useEffect(() => {
    setBreakpoint(getBreakpoint);
  }, [width]);

  flat.forEach((child, i) => {
    const colIndex = i % columns;

    cols[colIndex].push(
      // eslint-disable-next-line react/no-array-index-key
      <MasonryItem key={i} gutter={gutter}>
        {child}
      </MasonryItem>,
    );
  });

  return (
    <Row gutter={gutter}>
      {cols.map((_, i) => (
        <Col
          // eslint-disable-next-line react/no-array-index-key
          key={i}
          xs={calculateCols(xs)}
          sm={calculateCols(sm)}
          md={calculateCols(md)}
          lg={calculateCols(lg)}
        >
          {cols[i]}
        </Col>
      ))}
    </Row>
  );
};

export default memo(Masonry);
