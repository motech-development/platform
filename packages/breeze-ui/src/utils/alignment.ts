export enum Alignment {
  Centre = 'center',
  Left = 'left',
  Right = 'right',
}

type TAlignment = `${Alignment}`;

export default TAlignment;
