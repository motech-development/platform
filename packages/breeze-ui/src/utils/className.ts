import TSpacing from './spacing';
import TTheme from './theme';

export const className = (...name: string[]) => name.join(' ');

export const themeClass = (theme: TTheme, name: string) => {
  let mapping: string;

  switch (theme) {
    case 'danger':
      mapping = 'red';
      break;
    case 'secondary':
      mapping = 'gray';
      break;
    case 'success':
      mapping = 'green';
      break;
    case 'warning':
      mapping = 'yellow';
      break;
    default:
      mapping = 'blue';
  }

  return name.replace('{theme}', mapping);
};

export const spacingClass = (spacing: TSpacing, name: string) => {
  let mapping: string;

  switch (spacing) {
    case 'lg':
      mapping = '6';
      break;
    case 'md':
      mapping = '4';
      break;
    case 'sm':
      mapping = '2';
      break;
    default:
      mapping = '0';
  }

  return name.replace('{spacing}', mapping);
};
