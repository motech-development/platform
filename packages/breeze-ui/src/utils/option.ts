interface IOption {
  name: string;
  value: string;
}

export default IOption;
