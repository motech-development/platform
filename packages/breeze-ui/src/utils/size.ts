type TSize = 'lg' | 'md' | 'sm';

export default TSize;
