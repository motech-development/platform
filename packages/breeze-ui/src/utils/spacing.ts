export enum Spacing {
  Large = 'lg',
  Medium = 'md',
  None = 'none',
  Small = 'sm',
}

type TSpacing = `${Spacing}`;

export default TSpacing;
