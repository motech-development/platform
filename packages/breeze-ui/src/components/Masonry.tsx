import { FC, ReactNode } from 'react';
import useBreakpoint from '../hooks/useBreakpoint';
import { TColumn, TGap } from '../utils/grid';
import Col from './Col';
import Row from './Row';

const calculateCols = (cols: number) => Math.round(12 / cols) as TColumn;

export interface IMasonryProps {
  children: ReactNode | ReactNode[];
  gutter?: TGap;
  lg: TColumn;
  md: TColumn;
  sm: TColumn;
  xs: TColumn;
}

const Masonry: FC<IMasonryProps> = ({
  children,
  gutter = 0,
  lg,
  md,
  sm,
  xs,
}) => {
  const breakpoints = {
    lg,
    md,
    sm,
    xs,
  };
  const breakpoint = useBreakpoint();
  const columns = breakpoints[breakpoint];
  const cols: ReactNode[][] = Array.from(new Array(columns), () => []);
  const flat = Array.isArray(children) ? children.flat() : [children];

  flat.forEach((child, i) => {
    const colIndex = i % columns;

    cols[colIndex].push(
      // @tailwind: mb-0 mb-1 mb-2 mb-3 mb-4 mb-5
      // eslint-disable-next-line react/no-array-index-key
      <div className={`mb-${gutter}`} key={i}>
        {child}
      </div>,
    );
  });

  return (
    <Row gutter={gutter}>
      {cols.map((_, i) => (
        <Col
          // eslint-disable-next-line react/no-array-index-key
          key={i}
          xs={calculateCols(xs)}
          sm={calculateCols(sm)}
          md={calculateCols(md)}
          lg={calculateCols(lg)}
        >
          {cols[i]}
        </Col>
      ))}
    </Row>
  );
};

export default Masonry;
