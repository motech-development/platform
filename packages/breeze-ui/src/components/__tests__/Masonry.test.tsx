import { render } from '@testing-library/react';
import { TColumn } from '../../utils/grid';
import Masonry from '../Masonry';

describe('Masonry', () => {
  let xs: TColumn;
  let sm: TColumn;
  let md: TColumn;
  let lg: TColumn;

  beforeEach(() => {
    xs = 1;
    sm = 2;
    md = 3;
    lg = 4;
  });

  describe('xs viewport', () => {
    beforeEach(() => {
      Object.defineProperty(window, 'matchMedia', {
        value: jest.fn().mockImplementation((query) => ({
          addEventListener: jest.fn(),
          addListener: jest.fn(),
          dispatchEvent: jest.fn(),
          matches: false,
          media: query,
          onchange: null,
          removeEventListener: jest.fn(),
          removeListener: jest.fn(),
        })),
        writable: true,
      });
    });

    it('should output with a single child component', () => {
      const { container } = render(
        <Masonry xs={xs} sm={sm} md={md} lg={lg}>
          <div data-testid="child-1">Child 1</div>
        </Masonry>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should output with a multiple child components', () => {
      const { container } = render(
        <Masonry xs={xs} sm={sm} md={md} lg={lg}>
          <div data-testid="child-1">Child 1</div>
          <div data-testid="child-2">Child 2</div>
        </Masonry>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('sm viewport', () => {
    beforeEach(() => {
      Object.defineProperty(window, 'matchMedia', {
        value: jest.fn().mockImplementation((query) => ({
          addEventListener: jest.fn(),
          addListener: jest.fn(),
          dispatchEvent: jest.fn(),
          matches: query === '(min-width: 576px)',
          media: query,
          onchange: null,
          removeEventListener: jest.fn(),
          removeListener: jest.fn(),
        })),
        writable: true,
      });
    });

    it('should output with a single child component', () => {
      const { container } = render(
        <Masonry xs={xs} sm={sm} md={md} lg={lg}>
          <div data-testid="child-1">Child 1</div>
        </Masonry>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should output with a multiple child components', () => {
      const { container } = render(
        <Masonry xs={xs} sm={sm} md={md} lg={lg}>
          <div data-testid="child-1">Child 1</div>
          <div data-testid="child-2">Child 2</div>
        </Masonry>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('md viewport', () => {
    beforeEach(() => {
      Object.defineProperty(window, 'matchMedia', {
        value: jest.fn().mockImplementation((query) => ({
          addEventListener: jest.fn(),
          addListener: jest.fn(),
          dispatchEvent: jest.fn(),
          matches: query === '(min-width: 768px)',
          media: query,
          onchange: null,
          removeEventListener: jest.fn(),
          removeListener: jest.fn(),
        })),
        writable: true,
      });
    });

    it('should output with a single child component', () => {
      const { container } = render(
        <Masonry xs={xs} sm={sm} md={md} lg={lg}>
          <div data-testid="child-1">Child 1</div>
        </Masonry>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should output with a multiple child components', () => {
      const { container } = render(
        <Masonry xs={xs} sm={sm} md={md} lg={lg}>
          <div data-testid="child-1">Child 1</div>
          <div data-testid="child-2">Child 2</div>
        </Masonry>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('lg viewport', () => {
    beforeEach(() => {
      Object.defineProperty(window, 'matchMedia', {
        value: jest.fn().mockImplementation((query) => ({
          addEventListener: jest.fn(),
          addListener: jest.fn(),
          dispatchEvent: jest.fn(),
          matches: query === '(min-width: 992px)',
          media: query,
          onchange: null,
          removeEventListener: jest.fn(),
          removeListener: jest.fn(),
        })),
        writable: true,
      });
    });

    it('should output with a single child component', () => {
      const { container } = render(
        <Masonry xs={xs} sm={sm} md={md} lg={lg}>
          <div data-testid="child-1">Child 1</div>
        </Masonry>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });

    it('should output with a multiple child components', () => {
      const { container } = render(
        <Masonry xs={xs} sm={sm} md={md} lg={lg}>
          <div data-testid="child-1">Child 1</div>
          <div data-testid="child-2">Child 2</div>
        </Masonry>,
      );

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
