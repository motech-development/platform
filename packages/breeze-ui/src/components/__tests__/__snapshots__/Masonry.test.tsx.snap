// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Masonry lg viewport should output with a multiple child components 1`] = `
<div
  class="grid grid-cols-12 gap-0"
>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-1"
        >
          Child 1
        </div>
      </div>
    </div>
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-2"
        >
          Child 2
        </div>
      </div>
    </div>
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
</div>
`;

exports[`Masonry lg viewport should output with a single child component 1`] = `
<div
  class="grid grid-cols-12 gap-0"
>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-1"
        >
          Child 1
        </div>
      </div>
    </div>
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
</div>
`;

exports[`Masonry md viewport should output with a multiple child components 1`] = `
<div
  class="grid grid-cols-12 gap-0"
>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-1"
        >
          Child 1
        </div>
      </div>
    </div>
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-2"
        >
          Child 2
        </div>
      </div>
    </div>
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
</div>
`;

exports[`Masonry md viewport should output with a single child component 1`] = `
<div
  class="grid grid-cols-12 gap-0"
>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-1"
        >
          Child 1
        </div>
      </div>
    </div>
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
</div>
`;

exports[`Masonry sm viewport should output with a multiple child components 1`] = `
<div
  class="grid grid-cols-12 gap-0"
>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-1"
        >
          Child 1
        </div>
      </div>
    </div>
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-2"
        >
          Child 2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Masonry sm viewport should output with a single child component 1`] = `
<div
  class="grid grid-cols-12 gap-0"
>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-1"
        >
          Child 1
        </div>
      </div>
    </div>
  </div>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    />
  </div>
</div>
`;

exports[`Masonry xs viewport should output with a multiple child components 1`] = `
<div
  class="grid grid-cols-12 gap-0"
>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-1"
        >
          Child 1
        </div>
      </div>
      <div
        class="mb-0"
      >
        <div
          data-testid="child-2"
        >
          Child 2
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Masonry xs viewport should output with a single child component 1`] = `
<div
  class="grid grid-cols-12 gap-0"
>
  <div
    class="relative text-left col-span-12 sm:col-span-6 md:col-span-4 lg:col-span-3"
  >
    <div
      class=""
    >
      <div
        class="mb-0"
      >
        <div
          data-testid="child-1"
        >
          Child 1
        </div>
      </div>
    </div>
  </div>
</div>
`;
